src/k8s/pkg/k8s/setup/kubeconfigs.go

```diff
@@ -58,11 +67,87 @@ func InitKubeconfigs(ctx context.Context, state *state.State, ca *cert.CertKeyPa
 			return fmt.Errorf("could not generate auth token for %s: %w", config.username, err)
 		}
 
-		err = utils.GenerateKubeconfig(token, ca.CertPem, config.path)
+		err = generateKubeconfig(token, ca.CertPem, config.path, overwrites)
 		if err != nil {
 			return fmt.Errorf("failed to generate kubeconfig for %s: %w", config.username, err)
 		}
 	}
 
 	return nil
 }
+
+// generateX509Kubeconfig creates a kubeconfig file with the given x509 certificate and key data.
+func generateX509Kubeconfig(keyPem, certPem, caCertPem []byte, path string, overwrites *ExtraKubeConfigArgs) error {
+	ip, port, err := getApiServerConfig(overwrites)
+	if err != nil {
+		return fmt.Errorf("failed to get apiserver config: %w", err)
+	}
+	return utils.TemplateAndSave(snap.Instance().Path("k8s/config/kubeconfig-with-x509.tmpl"),
+		struct {
+			CaData        string
+			ApiServerIp   string
+			ApiServerPort int
+			CertData      string
+			KeyData       string
+		}{
+			CaData:        base64.StdEncoding.EncodeToString(caCertPem),
+			ApiServerIp:   ip,
+			ApiServerPort: port,
+			CertData:      base64.StdEncoding.EncodeToString(certPem),
+			KeyData:       base64.StdEncoding.EncodeToString(keyPem),
+		},
+		path,
+	)
+}
+
+// generateKubeconfig creates a kubeconfig file with the given token and CA data.
+func generateKubeconfig(token string, caCertPem []byte, path string, overwrites *ExtraKubeConfigArgs) error {
+	ip, port, err := getApiServerConfig(overwrites)
+	if err != nil {
+		return fmt.Errorf("failed to get apiserver config: %w", err)
+	}
+	return utils.TemplateAndSave(snap.Instance().Path("k8s/config/kubeconfig-with-token.tmpl"),
+		struct {
+			CaData        string
+			ApiServerIp   string
+			ApiServerPort int
+			Token         string
+		}{
+			CaData:        base64.StdEncoding.EncodeToString(caCertPem),
+			ApiServerIp:   ip,
+			ApiServerPort: port,
+			Token:         token,
+		},
+		path,
+	)
+}
+
+func getApiServerConfig(overwrites *ExtraKubeConfigArgs) (ip string, port int, err error) {
+	if overwrites == nil {
+		overwrites = &ExtraKubeConfigArgs{}
+	}
+
+	if overwrites.ApiServerIp != nil {
+		if net.ParseIP(*overwrites.ApiServerIp) == nil {
+			return "", 0, fmt.Errorf("%s is not a valid ip", *overwrites.ApiServerIp)
+		}
+		ip = *overwrites.ApiServerIp
+	} else {
+		ip = "127.0.0.1"
+	}
```

Comment:
Hostnames can also work here (e.g. `my-lb.example.com`). I would skip this check. Maybe set `ip` (or `host`) to `127.0.0.1`, or use the override.